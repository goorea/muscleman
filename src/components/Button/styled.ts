import styled, { css } from 'styled-components/native';

import { flexCenter } from '@src/styles/flex';
import { Colors } from '@src/types/theme';

export const StyledButton = styled.TouchableOpacity<{
  color?: keyof Colors;
  type?: 'solid' | 'clear' | 'outline';
  disabled?: boolean;
}>`
  ${({ type = 'solid', theme, color }) =>
    type === 'solid'
      ? css`
          background-color: ${theme[color || 'primary']};
        `
      : type === 'outline' &&
        css`
          border: 1px solid ${theme[color || 'primary']};
        `}

  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  padding: ${({ type }) => (type === 'clear' ? 0 : '14px 20px')};
  ${flexCenter};
`;
