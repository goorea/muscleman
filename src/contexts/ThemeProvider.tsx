import React from 'react';
import theme from '@src/theme';
import { useColorScheme } from 'react-native';
import { ThemeProvider as StyledThemeProvider } from 'styled-components/native';
import { ThemeContextState } from '@src/types/theme';

export const ThemeContext = React.createContext<ThemeContextState>({
  dark: false,
  toggleTheme: () => null,
  colors: theme.light,
});

const ThemeProvider: React.FC = ({ children }) => {
  const dark = useColorScheme() === 'dark';
  const [themeState, setTheme] = React.useState<ThemeContextState>({
    dark,
    toggleTheme: () => {
      setTheme(prevState => ({
        dark: !prevState.dark,
        toggleTheme: prevState.toggleTheme,
        colors: prevState.dark ? theme.light : theme.dark,
      }));
    },
    colors: dark ? theme.dark : theme.light,
  });

  React.useEffect(() => {
    if (dark !== themeState.dark) {
      themeState.toggleTheme();
    }
  }, [dark, themeState]);

  return (
    <ThemeContext.Provider value={themeState}>
      <StyledThemeProvider theme={themeState.colors}>
        {children}
      </StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

export function useTheme(): ThemeContextState {
  return React.useContext<ThemeContextState>(ThemeContext);
}

export default ThemeProvider;
