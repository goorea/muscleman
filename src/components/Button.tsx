import React from 'react';
import { ActivityIndicator, TouchableOpacityProps } from 'react-native';
import styled, { css } from 'styled-components/native';
import Text, { TextProps } from '@src/components/Text';
import { useTheme } from '@src/contexts/ThemeProvider';

type P = TouchableOpacityProps &
  TextProps & {
    title: string;
    type?: 'solid' | 'clear' | 'outline';
    loading?: boolean;
    disabled?: boolean;
  };

const RNTouchableOpacity = styled.TouchableOpacity<P>`
  ${({ type = 'solid', theme, color }) =>
    type === 'solid'
      ? css`
          background-color: ${theme[color || 'primary']};
        `
      : type === 'outline' &&
        css`
          border: 1px solid ${theme[color || 'primary']};
        `}

  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  padding: ${({ type }) => (type === 'clear' ? 0 : '6px 12px')};
`;

const Button: React.FC<P> = props => {
  const { colors } = useTheme();

  return (
    <RNTouchableOpacity {...props}>
      {props.loading ? (
        <ActivityIndicator
          size={props.size || 16}
          color={colors[props.color || 'primary']}
        />
      ) : (
        <Text
          weight={props.weight || 'bold'}
          color={
            props.type === undefined || props.type === 'solid'
              ? 'white'
              : props.color || 'primary'
          }>
          {props.title}
        </Text>
      )}
    </RNTouchableOpacity>
  );
};

export default Button;
