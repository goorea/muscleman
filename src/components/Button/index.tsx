import React from 'react';
import { ActivityIndicator, TouchableOpacityProps } from 'react-native';

import Icon, { IconProps } from '@src/components/Icon';
import Text, { TextProps } from '@src/components/Text';
import { useTheme } from '@src/contexts/ThemeProvider';
import { Colors } from '@src/types/theme';

import { StyledButton } from './styled';

type P = TouchableOpacityProps &
  Pick<TextProps, 'weight' | 'size'> & {
    title?: string | React.ReactElement;
    icon?: IconProps;
    type?: 'solid' | 'clear' | 'outline';
    titleColor?: keyof Colors;
    color?: keyof Colors;
    loading?: boolean;
    disabled?: boolean;
  };

const Button: React.FC<P> = props => {
  const {
    weight = 'bold',
    size = 16,
    title,
    icon,
    type = 'solid',
    titleColor = 'white',
    color = 'primary',
    loading,
  } = props;
  const { colors } = useTheme();

  return (
    <StyledButton {...props}>
      {loading ? (
        <ActivityIndicator size={size} color={colors[titleColor]} />
      ) : title ? (
        <Text
          weight={weight}
          color={type === undefined || type === 'solid' ? titleColor : color}>
          {title}
        </Text>
      ) : (
        icon && <Icon {...icon} />
      )}
    </StyledButton>
  );
};

export default Button;
