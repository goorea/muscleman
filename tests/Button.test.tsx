import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import ThemeProvider from '@src/contexts/ThemeProvider';
import Button, { ButtonProps } from '@src/components/Button';

describe('Button 컴포넌트', () => {
  const rendered = (props?: ButtonProps) =>
    render(
      <ThemeProvider>
        <Button title="버튼" {...props} />
      </ThemeProvider>,
    );

  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = rendered();

    expect(toJSON()).toMatchSnapshot();
  });
});
