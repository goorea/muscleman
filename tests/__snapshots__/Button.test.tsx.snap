// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 컴포넌트 렌더링이 올바르게 된다 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#2089dc",
      "opacity": 1,
      "paddingBottom": 6,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 6,
    }
  }
>
  <Text
    color="white"
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontFamily": "SpoqaHanSansNeo-Bold",
          "fontSize": 16,
        },
      ]
    }
    weight="bold"
  >
    버튼
  </Text>
</View>
`;
